python/src/equistore/operations/join.py

```diff
@@ -0,0 +1,209 @@
+             samples than the input TensorMap.
+    """
+
+    if len(tensormaps) < 2:
+        raise ValueError("provifde at least 2 tensormaps for joining")
+    for ts_to_join in tensormaps[1:]:
+        _check_same_keys(tensormaps[0], ts_to_join, "join")
+
+    blocks = []
+    for key in tensormaps[0].keys:
+        blocks_to_join = [ts.block(key) for ts in tensormaps]
+
+        if axis == "properties":
+            blocks.append(_join_blocks_along_properties(blocks_to_join))
+        elif axis == "samples":
+            blocks.append(_join_blocks_along_samples(blocks_to_join))
+        else:
+            raise ValueError(
+                "Only `'properties'` or `'samples'` are "
+                "valid values for the `axis` parameter."
+            )
+
+    return TensorMap(tensormaps[0].keys, blocks)
+
+
+def _join_labels(labels: List[Labels]) -> Labels:
+    """Join a sequence of :py:class:`Labels`
+
+    Possible name and value clashes are resolved by the following strategy:
+
+    1. If the names are the same and the values are unique we keep the names and
+       join the values vertically. In example::
+
+        labels_1 = ["n"], [0, 2, 3]
+        labels_2 = ["n"], [1, 4, 5]
+
+      will lead to::
+
+        properties = ["n"], [0, 2, 3, 1, 4, 5]
+
+    2. If the names are the same but the values are not unique a new variable
+       "tensor" is added to the names. In example::
+
+            properties_1 = ["n"], [0, 2, 3]
+            properties_2 = ["n"], [0, 2]
+
+        will lead to::
+            properties = ["tensor", "n"], [[0, 0], [0, 2], [0, 3], [1, 0], [1, 2]]
+
+    3. If names of the Labels are different we change the names two to
+       ("tensor", "property"). In example::
+
+            properties_1 = ["a"], [0, 2, 3]
+            properties_2 = ["b", "c], [[0, 0], [1, 2]]
+
+        will lead to::
+
+            properties = ["tensor", "q"], [[0, 0], [0, 1], [0, 2], [1, 0], [1, 1]]
+    """
+    names = [label.names for label in labels]
+    values = [label.tolist() for label in labels]
+    tensor_values = np.repeat(
+        a=np.arange(len(values)), repeats=[len(value) for value in values]
+    )
+
+    unique_names = np.unique(names)
+    unique_values = np.unique(np.vstack(values), axis=0)
+
+    names_are_same = np.all(
+        len(unique_names) == np.array([len(name) for name in names])
+    )
+    values_are_unqiue = len(unique_values) == len(np.vstack(values))
+
+    if names_are_same and values_are_unqiue:
+        new_names = names[0]
+        new_values = np.vstack(values)
+    elif names_are_same and not values_are_unqiue:
+        new_names = ["tensor"] + list(names[0])
+        new_values = np.hstack([tensor_values.reshape(-1, 1), np.vstack(values)])
+    else:
+        new_names = ["tensor", "property"]
+        property_values = np.hstack([np.arange(len(value)) for value in values])
+        new_values = np.vstack([tensor_values, property_values]).T
```

Comment:
I like the clarity of the logic here!

Reply:
Thanks. I tried my best that we still get this, if we don't look at the code for a while.